### PR TITLE
I refactored a bit. But it still needs work.

### PR DESCRIPTION
**Pull request by gojamesgo85**
We're now working with mainly one array, converting it into a set, and removing cocktails that don't match preferences instead of adding cocktails to new arrays. 

I can't figure out how to refactor the unwanted array portion. I'll need help with that one.

### CODE REVIEW

File: MetaCocktailsSwiftData/Views/Search View/SearchCriteriaViewModel.swift [hunk 1]
```diff
@@ -12,7 +12,7 @@ final class SearchCriteriaViewModel: ObservableObject {
     
     @Published var searchText: String = ""
     @Published var cocktailComponents = createComponentArray().sorted(by: { $0.name < $1.name })
```
Comment (MatthewHunt84) [inline MetaCocktailsSwiftData/Views/Search View/SearchCriteriaViewModel.swift:14]: Lets make a stored array here like we did for GeneratedBoozeComponents(), no reason to have to run a function to loop over every Item in the array each time we search. We should just run in the first time and store it.

**Review by MatthewHunt84 — CHANGES_REQUESTED** [review]
Okay, so pretty good but here’s the first thing we can improve. We use filter to create a preferredArray on line 122. This is a filtered array of COCKTAILCOMPONENTS that have .isPreferred set to true. 

Good so far.
Then, we start our matching process off in a pretty backwards way. We make five different arrays of any bases, styles, profiles, textures, or flavors that match what we have in our preferredArray of COCKTAILCOMPONENTS (or all of them count as matching if none have been selected). So not only do we have five different arrays, it's even worse because they're just an array of STRINGS

Why would we want five arrays of STRINGS?

We’re basically taking a CocktailComponent and throwing away all data that isn’t var name: String.  Look at CocktailComponent below

  class CocktailComponent: Identifiable, ObservableObject, Hashable {
    
    @Published var matchesCurrentSearch: Bool
    var id = UUID()
    var name: String
    @Published var isPreferred: Bool = false
    @Published var isUnwanted: Bool = false
    var isSpirit: Bool = false
    var isFlavor: Bool = false
    var isProfile: Bool = false
    var isStyle: Bool = false
    var isTexture: Bool = false
    var preferenceType: PreferenceType
    var spiritCategory: IngredientType?
    var spiritCategoryName: String = "" 
    
Look at all that information we could use to map and filter with that we’re throwing out the window! So first up lets refactor so that instead of 5 arrays of strings we finish with ONE array of COCKTAILCOMPONENTS.

This will break all the functionality below line 155 - so to do this lets make a new function called getFilteredCocktailsPart2() or something, and do this there. Then as we go we'll continue to build out this new refactoring function.

**Comment by gojamesgo85** [comment]
> Look at all that information we could use to map and filter with that we’re throwing out the window! So first up lets refactor so that instead of 5 arrays of strings we finish with ONE array of COCKTAILCOMPONENTS.


We already have that. Thats what preferredArray is. So we already have one array of CocktailComponents.

**Review by MatthewHunt84 — CHANGES_REQUESTED** [review]
Its looking great man! Nice job.

This code could be a little cleaner though. Before we talk on Sunday I want you to try a couple more things:

any time you've made a _for in_ loop, i.e:

`for _ in _ {...`

lets look at that code and see if there's a map / filter / reduce that we can use instead.
If there is, lets do that instead. If there isn't, or it's confusing, you can just leave it.

For example, 

                for booze in cocktail.spec {
                    if booze.ingredient.name == preferred.name {
                        matchCounter += 1
                    }
                }

seems like a good candidate for a _reduce_ function if you ask me. Take a look at each one, and consider if it can be re-written.
            
Secondly, 

    `filterUnwantedCocktails(cocktailComponentArray: [CocktailComponent], cocktails: [Cocktail]) -> [Cocktail]`

is a good idea for a function, but could be cleaner. Right now we have a **filter** function that has a **for in** loop inside it. It's just kinda messy to have the old way and the new way in the same method. Instead lets see if we can get real fancy and do two filter functions on the same array like this:

i.e:

    `cocktails.filter { remove cocktails that have unwanted preferences }
              .filter { second pass to remove cocktails with tags that match unwanted preferences }`

This one is harder so no worries if you get stuck. Good luck!

**Pull request merged**